Allow customizing retryable message types in Faulty agent tests

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37450 Allow customizing retryable message types in Faulty agent tests**

It doesn't seem like we could customize the retryable message types by
passing faulty_messages into dist_utils, as the `FaultyRpcAgentTestFixture`
overrode the `rpc_backend_options` function and provided the default list of
retryable message types. Needed to fix this as part of adding timeout injection
support as mentioned in https://github.com/pytorch/pytorch/issues/36272

Differential Revision: [D21270127](https://our.internmc.facebook.com/intern/diff/D21270127/)

rohan-varma committed:
Allow customizing retryable message types in Faulty agent tests

It doesn't seem like we could customize the retryable message types by
passing faulty_messages into dist_utils, as the `FaultyRpcAgentTestFixture`
overrode the `rpc_backend_options` function and provided the default list of
retryable message types. Needed to fix this as part of adding timeout injection
support as mentioned in https://github.com/pytorch/pytorch/issues/36272

Differential Revision: [D21270127](https://our.internmc.facebook.com/intern/diff/D21270127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow customizing retryable message types in Faulty agent tests"


It doesn't seem like we could customize the retryable message types by
passing faulty_messages into dist_utils, as the `FaultyRpcAgentTestFixture`
overrode the `rpc_backend_options` function and provided the default list of
retryable message types. Needed to fix this as part of adding timeout injection
support as mentioned in https://github.com/pytorch/pytorch/issues/36272

Differential Revision: [D21270127](https://our.internmc.facebook.com/intern/diff/D21270127/)

[ghstack-poisoned]

osalpekar commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -65,8 +65,12 @@ def new_test_method(self, *arg, **kwargs):
-        if faulty_messages:
-            _build_faulty_backend_options(faulty_messages)
+        if (
+            rpc.backend_registry.backend_registered("FAULTY_PROCESS_GROUP")
```
Do we need to check whether the backend is registered or not? The backend type is set by the faulty agent fixture, which imports `torch/distributed/rpc/_testing/__init__.py` to register the backend. The `__init__.py` raises a RuntimError if the backend is not available, so this check isn't necessary right?

rohan-varma replied:
Yes, an error would be raised, but some tests (i.e. `jit/rpc_spawn`) don't import the faulty agent fixture at all, so the backend will not be registered, and then check `self.rpc_backend
            == rpc.backend_registry.BackendType.FAULTY_PROCESS_GROUP` would throw during the test.

osalpekar reviewed (commented):
Left a few comments - thanks for the fixes!

rohan-varma committed:
Update on "Allow customizing retryable message types in Faulty agent tests"


It doesn't seem like we could customize the retryable message types by
passing faulty_messages into dist_utils, as the `FaultyRpcAgentTestFixture`
overrode the `rpc_backend_options` function and provided the default list of
retryable message types. Needed to fix this as part of adding timeout injection
support as mentioned in https://github.com/pytorch/pytorch/issues/36272

Differential Revision: [D21270127](https://our.internmc.facebook.com/intern/diff/D21270127/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -23,6 +23,17 @@ def _backend_type_repr(self):
+    Arguments:
+        backend_name (str): string to identify the RPC backend.
+    Returns:
+        True if the backend has been registered with `register_backend`, else False.
```
nit: double quote for `register_backend`

BTW, is this a user-facing API?

rohan-varma replied:
It's not intended to by used by users, and we don't export this function anywhere in the RPC docs.

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -96,19 +100,25 @@ def new_test_method(self, *arg, **kwargs):
+    messages_to_fail = (
+        faulty_messages
+        if faulty_messages is not None
+        else faulty_agent_fixture.retryable_message_types
```
thx for the fix!

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "Allow customizing retryable message types in Faulty agent tests"


It doesn't seem like we could customize the retryable message types by
passing faulty_messages into dist_utils, as the `FaultyRpcAgentTestFixture`
overrode the `rpc_backend_options` function and provided the default list of
retryable message types. Needed to fix this as part of adding timeout injection
support as mentioned in https://github.com/pytorch/pytorch/issues/36272

Differential Revision: [D21270127](https://our.internmc.facebook.com/intern/diff/D21270127/)

[ghstack-poisoned]
